exchange/bitswap/bitswap.go

```diff
@@ -270,26 +270,40 @@ func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg
 	// TODO: this is bad, and could be easily abused.
 	// Should only track *useful* messages in ledger
 
-	if len(incoming.Blocks()) == 0 {
+	iblocks := incoming.Blocks()
+
+	if len(iblocks) == 0 {
 		return
 	}
 
 	// quickly send out cancels, reduces chances of duplicate block receives
 	var keys []u.Key
-	for _, block := range incoming.Blocks() {
+	for _, block := range iblocks {
 		keys = append(keys, block.Key())
 	}
 	bs.wm.CancelWants(keys)
 
-	for _, block := range incoming.Blocks() {
+	for _, block := range iblocks {
 		bs.counterLk.Lock()
 		bs.blocksRecvd++
-		if has, err := bs.blockstore.Has(block.Key()); err == nil && has {
+		has, err := bs.blockstore.Has(block.Key())
```

Comment:
>  it makes the most sense to have a for{go waitgroup loop so no hasblock call has to wait on any other hasblock call.

yeah makes sense. though what about multiple calls to ReceiveMessage itself? they currently would wait on each other i think.